### DISCUSSION TITLE
CSS parsing using PEGTL?

### DISCUSSION THREAD

**Discussion by vibraphone** [taocpp/PEGTL#266] [Q&A] [answered]
Does a CSS parser using PEGTL already exist? If not, does anyone have recommendations on how to structure it? I'm especially curious about

1. how to detect and handle encodings (separate rulesets for utf-8, utf-16be/le, utf-32? or a new `tao::pegtl::xxx_input` object?)
2. how to separate tokenization from the grammar-parsing
3. how to deal with error handling/recovery.

**Comment by ColinH** [comment, marked as answer]
Hi,

there probably isn't a publicly available CSS grammar or parser built with the PEGTL, or at least I don't know about it. There was https://github.com/taocpp/PEGTL/discussions/184 but I also don't know if anything ever came of it.

1. We've never tried to solve the problem of using a single grammar to parse different Unicode encodings. What I would probably do to get started is (a) implement the grammar for UTF-8, (b) factor out the Unicode parts and "template over them" so that you can choose the Unicode encoding yielding 5 instantiations of the grammar, (c) come up with some auto-detect heuristic if there's not enough meta-data to decide which one to use.
2. That's not really the PEG approach which explicitly and intentionally works without a dedicated lexer stage. We have been doing some small steps in this direction (building a lexer with the PEGTL is straightforward, the part that still needs some work is parsing a sequence of tokens instead of a sequence of bytes. This part should become much easier/better once we get around to reworking the position tracking part of the input layer. For now there's only the limited `src/example/pegtl/token_input.cpp`.) I'd recommend not separating the two phases.
3. For error handling we usually use exceptions that are thrown with custom error messages in strategic places of the grammar, see for example `src/example/pegtl/json_errors.hpp` and check out where and how it is used. For error recovery we don't usually do much and fail at the first real error.

**Reply by vibraphone** [reply]
@ColinH Thanks for the reply. It's far from complete, but [here](https://github.com/vibraphone/pegtl-css/) is a first pass at a grammar that parses at least one non-trivial [example](https://kitware.github.io/vtk-js/css/navy.css). I haven't tried tackling anything but UTF-8 at this point. If you have any comments on particularly awful things I've done, I'd appreciate them.

**Reply by patlecat** [reply]
@vibraphone That's great, hope you can finish it, would be much needed to have a place where people can put in their grammars, like a vault or something. I'm looking for a handlebar or mustache grammar.
